incubator/binding-grpc/src/main/java/io/aklivity/zilla/runtime/binding/grpc/internal/config/GrpcOptionsConfigAdapter.java

```diff
@@ -0,0 +1,134 @@
+
+import io.aklivity.zilla.runtime.binding.grpc.internal.GrpcBinding;
+import io.aklivity.zilla.runtime.binding.grpc.internal.parser.Protobuf3Lexer;
+import io.aklivity.zilla.runtime.binding.grpc.internal.parser.Protobuf3Parser;
+import io.aklivity.zilla.runtime.engine.config.ConfigAdapterContext;
+import io.aklivity.zilla.runtime.engine.config.OptionsConfig;
+import io.aklivity.zilla.runtime.engine.config.OptionsConfigAdapterSpi;
+
+public final class GrpcOptionsConfigAdapter implements OptionsConfigAdapterSpi, JsonbAdapter<OptionsConfig, JsonObject>
+{
+    private static final String SERVICES_NAME = "services";
+    private Function<URL, String> readURL;
+    private ConfigAdapterContext context;
+
+    @Override
+    public Kind kind()
+    {
+        return Kind.BINDING;
+    }
+
+    @Override
+    public String type()
+    {
+        return GrpcBinding.NAME;
+    }
+
+    @Override
+    public JsonObject adaptToJson(
+        OptionsConfig options)
+    {
+        GrpcOptionsConfig grpcOptions = (GrpcOptionsConfig) options;
+
+        JsonObjectBuilder object = Json.createObjectBuilder();
+
+        if (grpcOptions.protobufConfigs != null)
+        {
+            JsonArrayBuilder keys = Json.createArrayBuilder();
+            grpcOptions.protobufConfigs.forEach(p -> keys.add(p.location));
+            object.add(SERVICES_NAME, keys);
+        }
+
+        return object.build();
+    }
+
+    @Override
+    public OptionsConfig adaptFromJson(
+        JsonObject object)
+    {
+        List<GrpcProtobufConfig> protobufConfigs = object.containsKey(SERVICES_NAME)
+                ? asListGrpcProtobufConfig(object.getJsonArray(SERVICES_NAME))
+                : null;
+
+        return new GrpcOptionsConfig(protobufConfigs);
+    }
+
+    @Override
+    public void adaptContext(
+        ConfigAdapterContext context)
+    {
+        this.readURL = context::readURL;
+    }
+
+    private List<GrpcProtobufConfig> asListGrpcProtobufConfig(
+        JsonArray array)
+    {
+        return array.stream()
+            .map(this::asGrpcProtobufConfig)
+            .collect(toList());
+    }
+
+    private GrpcProtobufConfig asGrpcProtobufConfig(
+        JsonValue value)
+    {
+        final String location = ((JsonString) value).getString();
+        final GrpcProtobufConfig protobufConfig = new GrpcProtobufConfig(location);
+        try
+        {
+            final String protoService = readURL.apply(new URL(location));
+            CharStream input = CharStreams.fromString(protoService);
+            Protobuf3Lexer lexer = new Protobuf3Lexer(input);
+            CommonTokenStream tokens = new CommonTokenStream(lexer);
+
+            Protobuf3Parser parser = new Protobuf3Parser(tokens);
+            parser.setErrorHandler(new BailErrorStrategy());
+            ParseTreeWalker walker = new ParseTreeWalker();
+            GrpcServiceDefinitionListener listener = new GrpcServiceDefinitionListener(protobufConfig);
```

Comment:
This should be a private inner class of this adapter class, with a shorter name. 😉